File: app/utils/watermark_logo.py
#!/usr/bin/env python3
#
# app/utils/watermark_logo.py
# Copyright (C) 2026 Gill-Bates http://github.com/Gill-Bates
#

"""Storage and validation for a user-supplied watermark logo.

The Settings page takes an SVG or a PNG, and what arrives here is always a PNG:
a dropped PNG is uploaded untouched, and a dropped SVG is rasterized on a
canvas in the browser and only the result is sent. Two reasons for that
asymmetry, both of them about what the server has to trust.

* **No SVG renderer on the server.** ffmpeg's librsvg decoder is present in
  some builds (Debian's) and absent in others - the static BtbN builds the
  Docker image ships from. Rasterizing here would work in development and fail
  in the image, which is the worst place to find out.
* **No SVG stored or served.** An SVG is a script-carrying document. Serving
  one back from our own origin would be a stored-XSS primitive; keeping only
  the flattened PNG removes the whole class. The browser's own rasterization
  is safe for the same reason: an SVG drawn through ``<img>`` runs in secure
  static mode - no scripts, no external references.

A PNG needs none of that: it is already flat pixels, so it travels as the user
exported it.

So the client-side checks are convenience, and everything in this module is the
actual boundary: it re-derives every property from the bytes it was handed,
using ffprobe/ffmpeg, and stores the file only once they all hold.
"""

from __future__ import annotations

import hashlib
import json
import logging
import subprocess
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Final

from .fs import get_data_dir

logger = logging.getLogger(__name__)

__all__ = [
    "LogoRejectedError",
    "LogoStatus",
    "custom_logo_path",
    "logo_status",
    "remove_custom_logo",
    "store_custom_logo",
]

# Its own subdirectory of the data volume, next to "cookies" and the badge
# cache: one named place per kind of thing an operator might mount, back up or
# inspect. Housekeeping only sweeps UUID-named job directories, so a plain name
# here is left alone.
_LOGO_DIRNAME: Final[str] = "logo"
_LOGO_FILENAME: Final[str] = "watermark-logo.png"

# 2 MiB is generous for logo artwork (the bundled one is ~10 KiB) and small
# enough that a rejected upload costs nothing.
MAX_LOGO_BYTES: Final[int] = 2 * 1024 * 1024

# Below the minimum the badge would be upscaled on any normal video; above the
# maximum there is nothing left to gain, since the badge is never drawn wider
# than 960px (watermark.py::_MAX_LOGO_WIDTH).
MIN_LOGO_PIXELS: Final[int] = 32
MAX_LOGO_PIXELS: Final[int] = 4096

# A badge sits in a corner. Wildly elongated artwork either shrinks to an
# illegible sliver or takes over the frame, so both extremes are refused.
MIN_LOGO_ASPECT: Final[float] = 0.1
MAX_LOGO_ASPECT: Final[float] = 20.0

# Pixel formats a PNG can decode to. Only the ones carrying an alpha channel
# are usable: a fully opaque logo would sit on the video as a solid rectangle.
# A palette PNG (pal8) is in here because its palette may carry transparency;
# whether it actually does is what _has_visible_transparency() then decides.
_ALPHA_PIXEL_FORMATS: Final[frozenset[str]] = frozenset({
    "rgba", "bgra", "argb", "abgr", "ya8", "ya16be", "ya16le",
    "rgba64be", "rgba64le", "pal8",
})

_PROBE_TIMEOUT_SECONDS: Final[int] = 20


class LogoRejectedError(Exception):
    """The uploaded file cannot be used as a watermark.

    The message is written for the person who dropped the file: it says what
    is wrong with their artwork, never what the server did to find out.
    """


@dataclass(frozen=True, slots=True)
class LogoStatus:
    """What the Settings page needs to render the drop zone."""

    custom: bool
    width: int = 0
    height: int = 0
    bytes: int = 0
    fingerprint: str = ""


def custom_logo_path() -> Path:
    return get_data_dir() / _LOGO_DIRNAME / _LOGO_FILENAME


def _probe_png(path: Path) -> tuple[int, int, str]:
    """``(width, height, pix_fmt)`` of a PNG, or raise LogoRejectedError."""
    cmd = [
        "ffprobe",
        "-v", "error",
        "-select_streams", "v:0",
        "-show_entries", "stream=codec_name,width,height,pix_fmt",
        "-of", "json",
        str(path),
    ]
    try:
        completed = subprocess.run(
            cmd, check=True, capture_output=True, timeout=_PROBE_TIMEOUT_SECONDS
        )
        stream = (json.loads(completed.stdout or "{}").get("streams") or [{}])[0]
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, json.JSONDecodeError, OSError) as exc:
        logger.info("Rejected watermark logo: not decodable (%s)", exc)
        raise LogoRejectedError("That file is not a readable image.") from exc

    if stream.get("codec_name") != "png":
        raise LogoRejectedError("That file is not a PNG image.")

    # ffprobe reports a stream for a file it only *guessed* was a PNG (it logs
    # the bad signature and exits 0), and gives it a zero size. So the size is
    # the check that a decoder actually got somewhere, not just a formality.
    width = int(stream.get("width") or 0)
    height = int(stream.get("height") or 0)
    if width <= 0 or height <= 0:
        raise LogoRejectedError("That file is not a readable image.")

    return width, height, str(stream.get("pix_fmt") or "")


def _has_visible_transparency(path: Path) -> bool:
    """Whether any pixel is not fully opaque.

    A logo without transparency is not artwork on the video, it is a box over
    it - so this is the one content check that decides usability rather than
    validity. Read straight off the alpha plane: fully opaque means the
    minimum alpha value is the maximum of the range.
    """
    cmd = [
        "ffmpeg",
        "-v", "error",
        "-i", str(path),
        # "format=rgba" first, so this holds for every PNG a user can export:
        # alphaextract has no plane to take from a palette image at all, and on
        # a 16-bit one it reports alpha on a 0-65535 scale that the comparison
        # below would read backwards. Both arrive here as 8-bit alpha.
        #
        # "file=-" puts the reading on stdout: metadata=print otherwise logs at
        # info level, which "-v error" swallows - and a silent probe would read
        # as "no transparency found" for every image.
        "-vf", "format=rgba,alphaextract,signalstats,metadata=print:key=lavfi.signalstats.YMIN:file=-",
        "-f", "null",
        "-",
    ]
    try:
        completed = subprocess.run(
            cmd, check=True, capture_output=True, timeout=_PROBE_TIMEOUT_SECONDS, text=True
        )
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, OSError) as exc:
        # An image whose alpha cannot be read is not worth failing an upload
        # over; the size and format checks already did the real work.
        logger.warning("Could not read the alpha channel of an uploaded logo: %s", exc)
        return True

    for line in (completed.stdout or "").splitlines():
        if "lavfi.signalstats.YMIN" in line:
            try:
                return float(line.rsplit("=", 1)[1]) < 255.0
            except (IndexError, ValueError):
                return True
    return True


def validate_logo_bytes(data: bytes, target: Path) -> LogoStatus:
    """Write ``data`` to ``target`` only if it is a usable watermark logo.

    Every check re-derives its answer from the bytes, never from what the
    client claimed. The file is written to a temporary name first, so a
    rejected upload cannot leave a half-valid logo behind.
    """
    if not data:
        raise LogoRejectedError("That file is empty.")
    if len(data) > MAX_LOGO_BYTES:
        raise LogoRejectedError(f"That image is larger than {MAX_LOGO_BYTES // (1024 * 1024)} MB.")

    target.parent.mkdir(parents=True, exist_ok=True)
    staged = target.with_name(f"{target.stem}.{uuid.uuid4().hex[:8]}.tmp.png")
    try:
        staged.write_bytes(data)

        width, height, pix_fmt = _probe_png(staged)
        if width < MIN_LOGO_PIXELS or height < MIN_LOGO_PIXELS:
            raise LogoRejectedError(
                f"That image is smaller than {MIN_LOGO_PIXELS}x{MIN_LOGO_PIXELS} pixels."
            )
        if width > MAX_LOGO_PIXELS or height > MAX_LOGO_PIXELS:
            raise LogoRejectedError(f"That image is larger than {MAX_LOGO_PIXELS} pixels on a side.")

        aspect = width / height
        if not MIN_LOGO_ASPECT <= aspect <= MAX_LOGO_ASPECT:
            raise LogoRejectedError("That image is too elongated to sit in a video corner.")

        if pix_fmt not in _ALPHA_PIXEL_FORMATS:
            raise LogoRejectedError(
                "That image has no transparency; it would cover the video with a solid box."
            )
        if not _has_visible_transparency(staged):
            raise LogoRejectedError(
                "Every pixel of that image is opaque; it would cover the video with a solid box."
            )

        staged.replace(target)
        return LogoStatus(
            custom=True,
            width=width,
            height=height,
            bytes=len(data),
            fingerprint=hashlib.sha256(data).hexdigest()[:16],
        )
    finally:
        staged.unlink(missing_ok=True)


def store_custom_logo(data: bytes) -> LogoStatus:
    """Validate and install an uploaded logo, replacing any previous one."""
    status = validate_logo_bytes(data, custom_logo_path())
    logger.info("Installed a custom watermark logo (%dx%d, %d bytes)", status.width, status.height, status.bytes)
    return status


def remove_custom_logo() -> bool:
    """Drop the custom logo and fall back to the bundled artwork."""
    path = custom_logo_path()
    try:
        path.unlink()
    except FileNotFoundError:
        return False
    except OSError as exc:
        logger.warning("Could not remove the custom watermark logo: %s", exc)
        raise
    logger.info("Removed the custom watermark logo")
    return True


def logo_status() -> LogoStatus:
    """Describe the logo currently in use, bundled or custom."""
    path = custom_logo_path()
    try:
        raw = path.read_bytes()
    except (FileNotFoundError, OSError):
        return LogoStatus(custom=False)

    try:
        width, height, _ = _probe_png(path)
    except LogoRejectedError:
        # Present but unreadable: report it as absent so the UI offers a fresh
        # upload rather than a broken preview. watermark.py ignores it too.
        logger.warning("The stored custom watermark logo is not readable; ignoring it")
        return LogoStatus(custom=False)

    return LogoStatus(
        custom=True,
        width=width,
        height=height,
        bytes=len(raw),
        fingerprint=hashlib.sha256(raw).hexdigest()[:16],
    )

File: app/static/js/watermark-logo.js
//
// app/static/js/watermark-logo.js
// Copyright (C) 2026 Gill-Bates http://github.com/Gill-Bates
//

// Prepares a dropped logo for upload. What leaves this file is always a PNG
// with an alpha channel - the only thing the server stores (see
// app/utils/watermark_logo.py).
//
// A PNG is passed through untouched: it is already flat pixels, so there is
// nothing to convert and nothing to re-encode.
//
// An SVG is rasterized here, in the browser, for two reasons the server cannot
// solve on its own: the static ffmpeg builds the Docker image ships have no SVG
// decoder, and an SVG kept on the server would be a script-carrying document we
// would eventually serve back. Drawing it through an <img> puts it in the
// browser's secure static mode - no scripts, no external references - and what
// comes out is flat pixels too.
//
// Everything below is convenience, not a security boundary: the server
// re-derives every property from the uploaded bytes and refuses the file if
// they do not hold.

/** Matches MAX_LOGO_BYTES in app/utils/watermark_logo.py. */
export const MAX_LOGO_BYTES = 2 * 1024 * 1024;

// Matches MIN/MAX_LOGO_ASPECT in app/utils/watermark_logo.py.
const MIN_ASPECT = 0.1;
const MAX_ASPECT = 20;

// Matches MIN/MAX_LOGO_PIXELS in app/utils/watermark_logo.py.
const MIN_LOGO_PIXELS = 32;
const MAX_LOGO_PIXELS = 4096;

// The badge is never drawn wider than 960px (watermark.py::_MAX_LOGO_WIDTH), so
// twice that leaves room for a clean lanczos downscale and nothing is gained
// above it. Capped at the server's MAX_LOGO_PIXELS on either side.
const RASTER_TARGET_WIDTH = 1920;

// An SVG that carries any of these is refused rather than quietly flattened:
// the browser would ignore them, but a logo that only looks right when its
// script runs is not the logo the user thinks they uploaded.
const FORBIDDEN_ELEMENTS = ["script", "foreignObject"];

/**
 * Structural check of an SVG's source text.
 *
 * @param {string} text
 * @returns {{ok: boolean, error?: string, width?: number, height?: number, aspect?: number}}
 */
export function inspectSvg(text) {
    const source = String(text || "").trim();
    if (!source) {
        return { ok: false, error: "That file is empty." };
    }

    let doc;
    try {
        doc = new DOMParser().parseFromString(source, "image/svg+xml");
    } catch {
        return { ok: false, error: "That file is not valid SVG." };
    }
    if (doc.querySelector("parsererror") || doc.documentElement?.localName !== "svg") {
        return { ok: false, error: "That file is not valid SVG." };
    }

    const root = doc.documentElement;

    for (const name of FORBIDDEN_ELEMENTS) {
        if (root.getElementsByTagName(name).length > 0) {
            return { ok: false, error: `That SVG contains <${name}>, which cannot be part of a watermark.` };
        }
    }

    for (const element of [root, ...root.querySelectorAll("*")]) {
        for (const attribute of Array.from(element.attributes || [])) {
            if (attribute.name.toLowerCase().startsWith("on")) {
                return { ok: false, error: "That SVG contains scripted event handlers." };
            }
            if (attribute.localName !== "href") {
                continue;
            }
            // Internal references (#gradient, #clip) are how normal artwork is
            // built; anything pointing outside the file would not load anyway.
            const value = String(attribute.value || "").trim();
            if (!value.startsWith("#")) {
                return { ok: false, error: "That SVG references external files, which will not render." };
            }
        }
    }

    const size = intrinsicSize(root);
    if (!size) {
        return { ok: false, error: "That SVG has no width/height or viewBox, so its size is unknown." };
    }
    const aspect = size.width / size.height;
    if (!(aspect >= MIN_ASPECT && aspect <= MAX_ASPECT)) {
        return { ok: false, error: "That SVG is too elongated to sit in a video corner." };
    }

    return { ok: true, width: size.width, height: size.height, aspect };
}

/** Intrinsic size from viewBox, falling back to width/height. */
function intrinsicSize(root) {
    const viewBox = String(root.getAttribute("viewBox") || "").trim();
    if (viewBox) {
        const parts = viewBox.split(/[\s,]+/).map(Number);
        if (parts.length === 4 && parts.every(Number.isFinite) && parts[2] > 0 && parts[3] > 0) {
            return { width: parts[2], height: parts[3] };
        }
    }

    // Only unitless or px values: "10em" depends on a font this document has no
    // context for, and guessing would silently distort the logo.
    const width = parseLength(root.getAttribute("width"));
    const height = parseLength(root.getAttribute("height"));
    return width && height ? { width, height } : null;
}

function parseLength(raw) {
    const match = /^\s*([0-9]*\.?[0-9]+)\s*(px)?\s*$/i.exec(String(raw || ""));
    if (!match) {
        return 0;
    }
    const value = Number(match[1]);
    return Number.isFinite(value) && value > 0 ? value : 0;
}

/** Raster dimensions for an aspect ratio, inside the server's pixel bounds. */
export function rasterSize(aspect) {
    let width = RASTER_TARGET_WIDTH;
    let height = Math.round(width / aspect);
    if (height > MAX_LOGO_PIXELS) {
        height = MAX_LOGO_PIXELS;
        width = Math.round(height * aspect);
    }
    if (width > MAX_LOGO_PIXELS) {
        width = MAX_LOGO_PIXELS;
        height = Math.round(width / aspect);
    }
    return { width: Math.max(1, width), height: Math.max(1, height) };
}

function loadImage(source, failure) {
    return new Promise((resolve, reject) => {
        const image = new Image();
        image.onload = () => resolve(image);
        image.onerror = () => reject(new Error(failure));
        image.src = source;
    });
}

/**
 * Render validated SVG text to a transparent PNG blob.
 *
 * @param {string} text
 * @param {number} aspect
 * @returns {Promise<Blob>}
 */
export async function rasterizeSvg(text, aspect) {
    const { width, height } = rasterSize(aspect);
    const url = URL.createObjectURL(new Blob([text], { type: "image/svg+xml" }));

    try {
        const image = await loadImage(url, "That SVG could not be rendered.");
        const canvas = document.createElement("canvas");
        canvas.width = width;
        canvas.height = height;

        const context = canvas.getContext("2d");
        if (!context) {
            throw new Error("This browser cannot render the logo.");
        }
        context.drawImage(image, 0, 0, width, height);

        assertUsableAlpha(context.getImageData(0, 0, width, height).data);

        const blob = await new Promise((resolve) => canvas.toBlob(resolve, "image/png"));
        if (!blob) {
            throw new Error("This browser could not encode the logo.");
        }
        return blob;
    } finally {
        URL.revokeObjectURL(url);
    }
}

/**
 * Which of the two accepted kinds a picked file claims to be, "" for neither.
 *
 * Kept in step with the drop zone's accept attribute in settings.html: a file
 * dragged onto the zone never passes through that filter.
 */
export function logoFileKind(file) {
    const name = String(file?.name || "");
    if (file?.type === "image/svg+xml" || /\.svg$/i.test(name)) {
        return "svg";
    }
    if (file?.type === "image/png" || /\.png$/i.test(name)) {
        return "png";
    }
    return "";
}

/**
 * The server's size and shape rules, stated for already-known dimensions.
 *
 * @returns {string} the reason the artwork is unusable, or "" if it is fine
 */
export function pixelSizeProblem(width, height) {
    if (!(width > 0 && height > 0)) {
        return "That file is not a readable image.";
    }
    if (width < MIN_LOGO_PIXELS || height < MIN_LOGO_PIXELS) {
        return `That image is smaller than ${MIN_LOGO_PIXELS}x${MIN_LOGO_PIXELS} pixels.`;
    }
    if (width > MAX_LOGO_PIXELS || height > MAX_LOGO_PIXELS) {
        return `That image is larger than ${MAX_LOGO_PIXELS} pixels on a side.`;
    }
    const aspect = width / height;
    if (!(aspect >= MIN_ASPECT && aspect <= MAX_ASPECT)) {
        return "That image is too elongated to sit in a video corner.";
    }
    return "";
}

/**
 * Check a dropped PNG's dimensions, decoding it only far enough to read them.
 *
 * The transparency check that the SVG path does on its own render is left to
 * the server here: it reads the real alpha plane with ffmpeg at full
 * resolution, where a browser would have to pull a multi-megabyte pixel buffer
 * back out of a canvas to reach the same answer less reliably.
 *
 * @param {File} file
 * @returns {Promise<{width: number, height: number}>}
 */
export async function inspectPng(file) {
    const url = URL.createObjectURL(file);
    try {
        const image = await loadImage(url, "That PNG could not be read.");
        const width = image.naturalWidth;
        const height = image.naturalHeight;
        const problem = pixelSizeProblem(width, height);
        if (problem) {
            throw new Error(problem);
        }
        return { width, height };
    } finally {
        URL.revokeObjectURL(url);
    }
}

/**
 * Turn a dropped file into the PNG blob the upload sends.
 *
 * @param {File} file
 * @returns {Promise<Blob>}
 */
export async function prepareLogoUpload(file) {
    if (!file) {
        throw new Error("No file to upload.");
    }
    if (file.size > MAX_LOGO_BYTES) {
        throw new Error(`That file is larger than ${MAX_LOGO_BYTES / (1024 * 1024)} MB.`);
    }

    const kind = logoFileKind(file);
    if (kind === "png") {
        // Uploaded byte for byte: re-encoding a PNG through a canvas would only
        // cost quality and size for a file that is already what we store.
        await inspectPng(file);
        return file;
    }
    if (kind === "svg") {
        const text = await file.text();
        const inspected = inspectSvg(text);
        if (!inspected.ok) {
            throw new Error(inspected.error);
        }
        return rasterizeSvg(text, inspected.aspect);
    }
    throw new Error("Please choose an SVG or a PNG file.");
}

/**
 * A watermark has to be artwork on the video, not a hole and not a box.
 *
 * @param {Uint8ClampedArray} pixels RGBA, as returned by getImageData
 */
export function assertUsableAlpha(pixels) {
    let hasInk = false;
    let hasTransparency = false;

    for (let index = 3; index < pixels.length; index += 4) {
        const alpha = pixels[index];
        if (alpha > 8) {
            hasInk = true;
        }
        if (alpha < 250) {
            hasTransparency = true;
        }
        if (hasInk && hasTransparency) {
            return;
        }
    }

    if (!hasInk) {
        throw new Error("That SVG renders as an empty image.");
    }
    throw new Error("That SVG has no transparency; it would cover the video with a solid box.");
}
